executor/src/lib.rs

```diff
@@ -278,7 +284,16 @@ impl DedicatedExecutor {
         let (tx, rx) = tokio::sync::oneshot::channel();
 
         let fut = Box::pin(async move {
-            let task_output = task.await;
+            let task_output = AssertUnwindSafe(task).catch_unwind().await.map_err(|e| {
+                if let Some(s) = e.downcast_ref::<String>() {
+                    s.clone()
+                } else if let Some(s) = e.downcast_ref::<&str>() {
+                    s.to_string()
+                } else {
+                    "unknown internal error".to_string()
```

Comment:
```suggestion
                    format!("unknown internal error: {}", e)
```

So we don't lose any contained message

Reply:
That doesn't work, let me explain why:

Panic payloads in Rust are `Any`, which neither implements `Display` nor `Debug`. Downcasting only works to concrete types, not to traits (e.g. `dyn Debug`). If you use the `panic!` macro (or `.expect(...)`), you get `&str` or `String`, depending if the macro contains formatting or not (see tests).

So what about the third case: I actually had to look up how to raise non-str/String payloads in Rust. This feature is mostly used for specialized applications like kernels or embedded systems, where you don't have enough resources for strings. While technically possible, nobody does that in any library I'm aware of. Also note that in this case, you still get a backtrace on stdout/stderr, you just don't get a nice error message. Note that the rust stdlib itself suffers from the same limitation. Hit "run" on this example and observe the error message: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=115230013145fc4f3c315e0969606a4a

Reply:
Thank you for the explanation -- I was remembering https://github.com/influxdata/influxdb_iox/pull/3130 where I had originally done something similar to here (special casing `&str` and `&String` and then found out later it was not necesary.

It sounds like this is not quite the same thing however